### PR TITLE
elide 3.0 createObject

### PR DESCRIPTION
**Pull request by tarrantzhang**
Implement new createObject interface in each datastore.

### CODE REVIEW

**Comment by yahoocla** [comment]
CLA is valid!

File: elide-core/src/main/java/com/yahoo/elide/core/PersistentResource.java [hunk 1]
```diff
@@ -124,9 +124,10 @@ public String toString() {
             Class<T> entityClass,
             RequestScope requestScope,
             String uuid) {
-        DataStoreTransaction tx = requestScope.getTransaction();
 
-        T obj = tx.createObject(entityClass);
+        //instead of calling transcation.createObject, create the new object here.
+        T obj = requestScope.getTransaction().createNewObject(entityClass);
```
Comment (aklish) [inline elide-core/src/main/java/com/yahoo/elide/core/PersistentResource.java]: Can't we just use reflection here directly to create the new instance?

File: elide-core/src/main/java/com/yahoo/elide/core/DataStoreTransaction.java [hunk 2]
```diff
@@ -71,15 +71,33 @@ default void preCommit() {
...
-     * @param entityClass the entity class
-     * @return new record
+     * @param entity   the persisted Object
+     * @param scope  the request scope
      */
-    <T> T createObject(Class<T> entityClass);
+    void createObject(Object entity, RequestScope scope);
 
+
+    default <T> T createNewObject(Class<T> entityClass) {
```
Comment (aklish) [inline elide-core/src/main/java/com/yahoo/elide/core/DataStoreTransaction.java]: Can we remove this method?

File: elide-core/src/main/java/com/yahoo/elide/core/DataStoreTransaction.java [hunk 3]
```diff
@@ -71,15 +71,33 @@ default void preCommit() {
...
     }
 
+//    /**
+//     * Create new entity record.
+//     *
+//     * @param <T>         the type parameter
+//     * @param entityClass the entity class
+//     * @return new record
+//     */
+//    <T> T createObject(Class<T> entityClass);
```
Comment (aklish) [inline elide-core/src/main/java/com/yahoo/elide/core/DataStoreTransaction.java]: We can remove this dead code.

File: elide-datastore/elide-datastore-inmemorydb/src/main/java/com/yahoo/elide/datastores/inmemory/InMemoryTransaction.java [hunk 4]
```diff
@@ -98,21 +100,16 @@ public void commit() {
...
-        try {
-            T instance = entityClass.newInstance();
-            setId(instance, id);
-            return instance;
-        } catch (InstantiationException | IllegalAccessException e) {
-            e.printStackTrace();
-        }
-        return null;
+        setId(entity, id);
+        operations.add(new Operation(id, entity, entity.getClass(), false));
```
Comment (aklish) [inline elide-datastore/elide-datastore-inmemorydb/src/main/java/com/yahoo/elide/datastores/inmemory/InMemoryTransaction.java]: What is this doing?

File: elide-core/src/main/java/com/yahoo/elide/core/PersistentResource.java (hunk 1 above)
Comment (wcekan) [inline elide-core/src/main/java/com/yahoo/elide/core/PersistentResource.java]: `T obj = (T) entityClass.newInstance();`

File: elide-datastore/elide-datastore-inmemorydb/src/main/java/com/yahoo/elide/datastores/inmemory/InMemoryTransaction.java (hunk 4 above)
Comment (tarrantzhang) [inline elide-datastore/elide-datastore-inmemorydb/src/main/java/com/yahoo/elide/datastores/inmemory/InMemoryTransaction.java]: operations holds all changes made, and will execute each operation during commit. createObject() used to follow by save() which does this job, but now we don't call save() after it anymore (we create with populated object, and then commit directly). So we need to add this to operations.

File: elide-datastore/elide-datastore-inmemorydb/src/main/java/com/yahoo/elide/datastores/inmemory/InMemoryTransaction.java [hunk 5]
```diff
@@ -57,7 +58,8 @@ public void save(Object object) {
         }
         String id = dictionary.getId(object);
         if (id.equals("0")) {
-            setId(object, dictionary.getId(createObject(object.getClass())));
+            createObject(object, null);
```
Comment (aklish) [inline elide-datastore/elide-datastore-inmemorydb/src/main/java/com/yahoo/elide/datastores/inmemory/InMemoryTransaction.java]: Let's see if this is still needed.  Let's throw an exception if id ever equals 0.

**Pull request merged**